python/cuml/common/array.py

```diff
@@ -0,0 +1,262 @@
+            ary_interface = data.__cuda_array_interface__
+
+        else:
+            raise TypeError("Unrecognized data type.")
+
+        if ary_interface:
+            super(Array, self).__init__(data=data, owner=owner)
+            self.shape = ary_interface['shape']
+            self.dtype = np.dtype(data.dtype)
+            if ary_interface['strides'] is None:
+                self.order = 'C'
+                self.strides = _order_to_strides(self.order, self.shape,
+                                                 self.dtype)
+            else:
+                self.strides = ary_interface['strides']
+                self.order = _strides_to_order(self.strides, data.dtype)
+
+    def __getitem__(self, slice):
+        return Array(data=cp.asarray(self).__getitem__(slice))
+
+    def __setitem__(self, slice, value):
+        cp.asarray(self).__setitem__(slice, value)
+
+    def __reduce__(self):
+        return self.__class__, (self.to_output('numpy'),)
+
+    def __len__(self):
+        return self.shape[0]
+
+    @property
+    def __cuda_array_interface__(self):
+        output = {
+            "shape": self.shape,
+            "strides": self.strides,
+            "typestr": self.dtype.str,
+            "data": (self.ptr, False),
+            "version": 2,
+        }
+        return output
+
+    def to_output(self, output_type='cupy'):
+        """
+        Convert array to output format
+
+        Parameters
+        ----------
+        output_type : string
+            Format to convert the array to. Acceptable formats are:
+            'cupy' - to cupy array
+            'numpy' - to numpy (host ) array
+            'numba' - to numba device array
+            'dataframe' - to cuDF DataFrame
+            'series' - to cuDF Series
+            'cudf' - to cuDF Series if array in single dimensional, to
+                DataFrame otherwise
+        """
+
+        if output_type == 'cudf':
+            if len(self.shape) == 1:
+                output_type = 'series'
+            elif self.shape[0] > 1 and self.shape[1] > 1:
+                output_type = 'dataframe'
+            else:
+                output_type = 'series'
+
+        if output_type == 'cupy':
+            return cp.asarray(self)
+
+        elif output_type == 'numba':
+            return cuda.as_cuda_array(self)
+
+        elif output_type == 'numpy':
+            return cp.asnumpy(cp.asarray(self), order=self.order)
+
+        elif output_type == 'dataframe':
+            if self.dtype not in [np.uint8, np.uint16, np.uint32,
+                                  np.uint64, np.float16]:
+                mat = cuda.as_cuda_array(self)
+                if len(mat.shape) == 1:
+                    mat = mat.reshape(mat.shape[0], 1)
+                return DataFrame.from_gpu_matrix(mat)
+            else:
+                raise ValueError('cuDF unsupported Array dtype')
+
+        elif output_type == 'series':
+            if len(self.shape) == 1:
```

Comment:
can this support be handled by the `cudf` branch of the if above, just adding `assert (single_dimensional) if output_type == series`? Otherwise seems like kind of repeated code